lib/ui/proposals/list/interactor/proposals_bloc.dart

```diff
@@ -1,68 +1,94 @@
     on<_Initial>(_initial);
+    on<_Load>(_loadProposalsForEvent);
   }
 
+  List<DaoData> daos = [];
   FilterStatus filterStatus = FilterStatus.active;
+  /// This value can be modified based on your preferences
+  final int first = 10;
+  late int offset;
 
   Future<void> _initial(_Initial event, Emitter<ProposalsState> emit) async {
     if (!event.refresh) {
       emit(state.copyWith(pageState: PageState.loading));
       filterStatus = event.filterStatus;
     }
 
-    await _fetchAndEmitProposals(emit, filterStatus);
+    await _fetchDaos(event, emit);
   }
 
-  Future<void> _fetchAndEmitProposals(Emitter<ProposalsState> emit, FilterStatus filterStatus) async {
-    // Fetch DAOs
+  Future<void> _fetchDaos(_Initial event, Emitter<ProposalsState> emit) async {
     final Result<ProfileData, HyphaError> profileResult = await _fetchProfileUseCase.run();
 
     if (profileResult.isValue && profileResult.asValue!.value.daos.isNotEmpty) {
       daos = profileResult.asValue!.value.daos;
-      // Fetch Proposals using the fetched DAOs
-      final Result<List<ProposalModel>, HyphaError> proposalsResult =
-          await _getProposalsUseCase.run(GetProposalsUseCaseInput(daos, filterStatus));
-
-      if (proposalsResult.isValue) {
-        // Emit both daos and proposals in one state
-        emit(state.copyWith(
-          pageState: PageState.success,
-          proposals: proposalsResult.asValue!.value,
-        ));
+
+      await _loadProposalsForEvent(event, emit);
+    } else {
+      final HyphaError error = profileResult.isError ? profileResult.asError!.error : HyphaError.api('Failed to retrieve DAOs');
+      await _errorHandlerManager.handlerError(error);
+      emit(state.copyWith(pageState: PageState.failure));
+    }
+  }
+
+  Future<void> _loadProposalsForEvent(dynamic event, Emitter<ProposalsState> emit) async {
+    offset = event is _Initial ? 0 : offset + first;
+
+    final List<DaoData> selectedDaos = (GetIt.instance<FilterProposalsBloc>().selectedDaoIds != null && event is _Load)
+        ? daos.where((dao) => GetIt.instance<FilterProposalsBloc>().selectedDaoIds!.contains(dao.docId)).toList()
+        : daos;
+
+    await _fetchAndEmitProposals(event, emit, selectedDaos, 0);
+  }
+
+  Future<void> _fetchAndEmitProposals(dynamic event, Emitter<ProposalsState> emit, List<DaoData> selectedDaos, int nb) async {
+    final Result<List<ProposalModel>, HyphaError> proposalsResult = await _getProposalsUseCase.run(GetProposalsUseCaseInput(selectedDaos, filterStatus, first, offset));
+
+    if (proposalsResult.isValue) {
+      final List<ProposalModel> proposals = event is _Initial ? proposalsResult.asValue!.value : state.proposals + proposalsResult.asValue!.value;
+
+      final List<int> daoIds = GetIt.instance<FilterProposalsBloc>().selectedDaoIds ?? daos.map((DaoData dao) => dao.docId).toList();
+      if (nb < 3 && proposalsResult.asValue!.value.filterByDao(daoIds).isEmpty) {
```

Comment:
what is the "nb" magic number - this needs to have a real name.

Reply:
You're right. I think `retryCounter` is a great choice because, in this case, the query can return an empty list or the filtering that happens after the query, even though there are more proposals on the next pages. So, if the initial list is empty (and we cannot scroll), the function calls itself up to a maximum of three times. This logic applies not only at the start of the list but also in the middle and the end.